lib/DependencyProvider.js

```diff
@@ -0,0 +1,317 @@
+// @flow
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const request = require('sync-request').default;
+
+// Cache of existing versions according to the package website.
+let onlineVersionsCache /*: Map<string, Array<string>> */ = new Map();
+
+// Memoization cache to avoid doing the same work twice in listAvailableVersions.
+// This is to be cleared before each call to solve_deps().
+const listVersionsMemoCache /*: Map<string, Array<string>> */ = new Map();
```

Comment:
We already have a cache in Solve.js. Can you explain how this cache comes into play?

Reply:
I would like to architect this such that:

1. `DependancyProvider` is a class
2. We create a new `DependancyProvider` for each call to `solve_deps()` rather than clearing a global variable.
3. We pass (bound?) methods from the `DependancyProvider` class to 

```
    const dependancyProvider = new DependencyProvider();
    
    // Solve again, in online mode.
    solution = wasm.solve_deps(
      JSON.stringify(elmJson),
      useTest,
      extra,
      () => dependancyProvider.fetchElmJsonOnline(),
      () => dependencyProvider.listAvailableVersionsOnline(),
    );
```

Reply:
(Note that last time I tried to use a class, I couldn’t make Flow and Prettier work with it.)

Reply:
One remark, the cache for memoization should be reset between calls, because otherwise you can't take into account updates of existing versions. But the cache for online versions doesn't need to be reset between solver calls, just updated (if we want) with `updateOnlineVersionsCache`. This can be a bit faster in the context of watch mode with changes in the `elm.json` invalidating the checksum cache. Not a lot faster, but a little bit, since it avoids the JSON file read and de-serialization into a hashmap. And just need to do the hashmap -> json -> write updated file serialization after the update is done.

Reply:
What do you think of something like this:
```js
  try {
    const offlineDepsProvider = new DependencyProvider.Offline();
    solution = offlineDepsProvider.solve(elmJson, useTest, extra);
  } catch (_) {
    // console.error('Offline solver failed, switching to new online');
    const syncGetWorker = SyncGet.startWorker();
    const onlineDepsProvider = new DependencyProvider.Online(syncGetWorker);
    solution = onlineDepsProvider.solve(elmJson, useTest, extra);
    syncGetWorker.shutDown();
  }
```
So inside the `Online` and `Offline` classes of the `DependencyProvider` module, we expose a `solve()` function that internally does the clearing of the memoization cache and uses the correct versions of `fetchElmJson` and `listAvailableVersions`. The online versions cache could be updated at the beginning of the `solve()` call, or only at the constructor initialization and then exposing an update function for the cache in the online class.

Here I'm also passing the object enabling synchronous GET requests based on a worker discussed in another thread here. This is because it's better to create the worker only once instead of one time per solve call, to avoid doing it in watch mode. But for simplicity, we could also do this directly inside the `Online` class. I don't think the perf change would be big anyway cause starting a worker shouldn't be too expensive I think. Especially considering that the tests runner already creates new child processes for the runners.